plugins/module_utils/compose_v2.py

```diff
@@ -348,6 +349,69 @@ def _concat_event_msg(event, append_msg):
     )
 
 
+def parse_json_events(stderr, warn_function=None):
+    events = []
+    stderr_lines = stderr.splitlines()
+    if stderr_lines and stderr_lines[-1] == b'':
+        del stderr_lines[-1]
+    for line in stderr_lines:
+        line = line.strip()
+        if not line.startswith(b'{') or not line.endswith(b'}'):
+            continue
+        try:
+            line_data = json.loads(line)
+        except Exception as exc:
+            if warn_function:
+                warn_function(
+                    'Cannot parse event from line: {0!r}: {1}. Please report this at '
+                    'https://github.com/ansible-collections/community.docker/issues/new?assignees=&labels=&projects=&template=bug_report.md'
+                    .format(line, exc)
+                )
+            continue
+        if line_data.get('error'):
+            resource_type = ResourceType.UNKNOWN
+            event = Event(
+                resource_type,
+                line_data.get('id'),
+                'Error',
+                line_data.get('message'),
+            )
+        else:
+            resource_type = ResourceType.UNKNOWN
+            resource_id = line_data.get('id')
+            status = line_data.get('status')
+            text = line_data.get('text')
+            if isinstance(resource_id, str) and ' ' in resource_id:
+                resource_type_str, resource_id = resource_id.split(' ', 1)
+                try:
+                    resource_type = ResourceType.from_docker_compose_event(resource_type_str)
+                except KeyError:
+                    if warn_function:
+                        warn_function(
+                            'Unknown resource type {0!r}. Please report this at '
+                            'https://github.com/ansible-collections/community.docker/issues/new?assignees=&labels=&projects=&template=bug_report.md'
+                            .format(resource_type_str)
+                        )
+                    resource_type = ResourceType.UNKNOWN
+            elif text in DOCKER_STATUS_PULL:
+                resource_type = ResourceType.IMAGE
+                status, text = text, None
+            elif text in DOCKER_PULL_PROGRESS_DONE or line_data.get('text') in DOCKER_PULL_PROGRESS_WORKING:
+                resource_type = ResourceType.IMAGE_LAYER
+                status, text = text, None
+            if status not in DOCKER_STATUS and text in DOCKER_STATUS:
+                status, text = text, status
+            event = Event(
+                resource_type,
+                resource_id,
+                status,
+                text,
+            )
+
```

Comment:
The structured output is still somewhat messy, but that's unfortunately not a surprise if you look at how the events are created that the JSON progress dumps (the progress dumping itself is rather dumb, it does not do fancy transforms).

If someone ever wants to rewrite Compose again, I would suggest to start with a proper machine readable output before trying to make it look nice. That will avoid quite some messy progess events from the beginning...